upstairs/src/lib.rs

```diff
@@ -3982,19 +3982,24 @@ impl Downstairs {
 
         if !response.block_contexts.is_empty() {
             let mut successful_hash = false;
-
             let computed_hash = integrity_hash(&[&response.data[..]]);
 
             // The most recent hash is probably going to be the right one.
             for context in response.block_contexts.iter().rev() {
                 if computed_hash == context.hash {
                     successful_hash = true;
-                    valid_hash = Some(context.hash);
                     break;
                 }
             }
 
-            if !successful_hash {
+            if successful_hash {
+                // Filter out contexts that don't match, and return the
+                // successful hash.
+                response
+                    .block_contexts
+                    .retain(|context| context.hash == computed_hash);
+                valid_hash = Some(computed_hash);
```

Comment:
Nit: is `valid_hash` even needed?  It seems like you could write this conditional as
```suggestion
                // Filter out contexts that don't match, and return the
                // successful hash.
                response
                    .block_contexts
                    .retain(|context| context.hash == computed_hash);
                Some(computed_hash)
```
then remove the `Ok(valid_hash)` at the end of the function (deleting `valid_hash` entirely).

Reply:
I think so but only when validating encrypted block contexts, where each `computed_hash` will be different. In the unencrypted case `computed_hash` is computed once and checked for each context.

Reply:
This function is called `validate_unencrypted_read_response`, so I don't think the encrypted case matters here 😄

Reply:
AH I see what you mean now! e4dd3fb